cmd/vcluster/cmd/hostpaths.go

```diff
@@ -0,0 +1,512 @@
+	options.VirtualLogsPath = filepath.Join(virtualPath, "log")
+	options.VirtualPodLogsPath = filepath.Join(options.VirtualLogsPath, "pods")
+
+	options.VirtualContainerLogsPath = filepath.Join(options.VirtualLogsPath, "containers")
+
+	err := os.Mkdir(options.VirtualContainerLogsPath, os.ModeDir)
+	if err != nil {
+		if !os.IsExist(err) {
+			klog.Errorf("error creating container dir in log path: %v", err)
+			return err
+		}
+	}
+
+	inClusterConfig := ctrl.GetConfigOrDie()
+
+	inClusterConfig.QPS = 40
+	inClusterConfig.Burst = 80
+	inClusterConfig.Timeout = 0
+
+	translate.Suffix = options.Name
+
+	var virtualClusterConfig *rest.Config
+	err = wait.PollImmediate(time.Second, time.Hour, func() (bool, error) {
+		virtualClusterConfig = &rest.Config{
+			Host: options.Name,
+			TLSClientConfig: rest.TLSClientConfig{
+				ServerName: options.Name,
+				CertFile:   options.ClientCaCert,
+				KeyFile:    options.ServerCaKey,
+				CAFile:     options.ServerCaCert,
+			},
+		}
+
+		kubeClient, err := kubernetes.NewForConfig(virtualClusterConfig)
+		if err != nil {
+			return false, errors.Wrap(err, "create kube client")
+		}
+
+		_, err = kubeClient.Discovery().ServerVersion()
+		if err != nil {
+			klog.Infof("couldn't retrieve virtual cluster version (%v), will retry in 1 seconds", err)
+			return false, nil
+		}
+		_, err = kubeClient.CoreV1().ServiceAccounts("default").Get(context.Background(), "default", metav1.GetOptions{})
+		if err != nil {
+			klog.Infof("default ServiceAccount is not available yet, will retry in 1 seconds")
+			return false, nil
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	localManager, err := ctrl.NewManager(inClusterConfig, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		LeaderElection:     false,
+		Namespace:          options.TargetNamespace,
+		NewClient:          pluginhookclient.NewPhysicalPluginClientFactory(blockingcacheclient.NewCacheClient),
+	})
+	if err != nil {
+		return err
+	}
+
+	virtualClusterManager, err := ctrl.NewManager(virtualClusterConfig, ctrl.Options{
+		Scheme:             scheme,
+		MetricsBindAddress: "0",
+		LeaderElection:     false,
+		NewClient:          pluginhookclient.NewVirtualPluginClientFactory(blockingcacheclient.NewCacheClient),
+	})
+	if err != nil {
+		return err
+	}
+
+	ctx := context.WithValue(context.Background(), optionsKey, options)
+
+	startManagers(ctx, localManager, virtualClusterManager)
+
+	mapHostPaths(ctx, localManager, virtualClusterManager)
+
+	return nil
+}
+
+func mapHostPaths(ctx context.Context, pManager, vManager manager.Manager) {
+	options := ctx.Value(optionsKey).(*context2.VirtualClusterOptions)
+
+	wait.Forever(func() {
+		podList := &corev1.PodList{}
+		err := pManager.GetClient().List(ctx, podList, &client.ListOptions{
+			Namespace: options.TargetNamespace,
+			FieldSelector: fields.SelectorFromSet(fields.Set{
+				NodeIndexName: os.Getenv(HostpathMapperSelfNodeNameEnvVar),
+			}),
+		})
```

Comment:
Additional small optimization here could be to use label selector to filter out pods in the namespace that are not synced by the particular vcluster instance based on the ["vcluster.loft.sh/managed-by"](https://github.com/loft-sh/vcluster/blob/e6f9bf2ddaf467869b24e51ac70e7e8bed942073/pkg/util/translate/translate.go#L18) label.
But this only makes an impact if there are multiple big vclusters in the same namespace, which is unlikely.